### ISSUE TITLE
Review open issues

### ISSUE THREAD

**Issue by smiths** [JacquesCarette/Drasil#2096]
Read through ALL of the open issues and create a list of the top 10 that interest you and "make sense".  You can respond with your list below.  In your list, use the hash symbol to reference the issues.  (For instance, the last issue I assigned you was #2095.)

When you have completed this task, you can check your name off:

- [x] Nathaniel @Nathaniel-Hu 
- [x] Muhammad @muhammadaliog3

**Comment by muhammadaliog3** [comment]
[#901](https://github.com/JacquesCarette/Drasil/issues/901)
[#872](https://github.com/JacquesCarette/Drasil/issues/872)
[#1048](https://github.com/JacquesCarette/Drasil/issues/1048)
[#806](https://github.com/JacquesCarette/Drasil/issues/806)

Contraints in general, this would solve a group of issues 
[#1799](https://github.com/JacquesCarette/Drasil/issues/1799)
[#1798](https://github.com/JacquesCarette/Drasil/issues/1798)
[#1795](https://github.com/JacquesCarette/Drasil/issues/1795)
[#1797](https://github.com/JacquesCarette/Drasil/issues/1797)

[#1476](https://github.com/JacquesCarette/Drasil/issues/1476)
[#1570](https://github.com/JacquesCarette/Drasil/issues/1570)
[#1809](https://github.com/JacquesCarette/Drasil/issues/1809)
[#1153](https://github.com/JacquesCarette/Drasil/issues/1153)
[#1732 ](https://github.com/JacquesCarette/Drasil/issues/1732)

**Comment by smiths** [comment]
@muhammadaliog3, can you provide more information on why the particular issues interest you?  You can edit your above comment, you don't need to create a new one.

**Comment by Nathaniel-Hu** [comment]
The following top 10 issues (not necessarily in order of decreasing interest) are interesting to me and "make sense":

1. #946 Determine Document Outline from Actual Document Contents
- would help me learn how to use Haskell to analyze documents and extract data from them, as well as process the extracted data to generate the document outline

2. #806 Sort References by Order of Appearance in Document
- very similar to above (Issue #946)

3. #801 Automatically generate traceability graphs
- similar to above two (Issues #946 and #806), would also help me become more familiarized with the traceability matrices/understand how traceability works

4. #1018 HTML Table of Contents
 - would help me become more familiar with the SRS formatting, the organization of information, etc.; would also allow me to practice learning/using some HTML

5. #2088 Generate design log with code
- would help me become more familiar with the algorithms/decision-making/reasoning used in Drasil's code generator

6. #1998 Implement linked lists in GOOL
- would help me learn more about data structures and how to create/implement them in Haskell

7. #1973 Run linters on generated code
- would help me learn more about code testing/evaluation/verification using code linters

8. #1916 Detect and complain if the same symbol is used for two or more concepts within the same document
- would help me learn how to use Haskell for document analysis

9. #1879 GOOL should have different kinds of variables
- would help me to better understand the variable domains/spaces that are designated in the Drasil framework (for global variables, local variables and external variables)

10. #1504 Automatically collect constants used for constraint bounds
- would help me learn how to use Haskell to analyze variable values for constant constraints vs. variable constraints

**Comment by smiths** [comment]
@JacquesCarette is in a better position to prioritize your list of ideas @Nathaniel-Hu, but in case you are looking for something to do, I suggest that you start with #801.  It is a well-defined task and something that doesn't rely too deeply on the Drasil infrastructure.  The Traceability Matrix is already generated.  We just want a more graphical view of this same information.  As @JacquesCarette in the original issue, you want to start with looking into generating a `.dot` file.

**Comment by smiths** [comment]
@muhammadaliog3, the best person to prioritize your list is @JacquesCarette.  However, I can get you started.  I agree with your recommendation of issue #1570.  It seems like a nice straightforward task.  I'm not sure why the original issue was never closed?

**Comment by Nathaniel-Hu** [comment]
@smiths where can I find the generated Traceability Matrices (on GitHub) for the Drasil infrastructure?

**Comment by smiths** [comment]
@Nathaniel-Hu, for questions about issue #801, you should ask them under the original issue, rather than here.  The generated traceability matrices are in the SRS documentation.  You can find them in one of the last sections.  For instance, you can look at the matrix for the Projectile example:

https://jacquescarette.github.io/Drasil/examples/Projectile/srs/Projectile_SRS.html

I do not know what specific Drasil code generates the matrix.  I'm optimistic that you can track it down, and that you will learn more about Drasil in the process.  :smile:  If you get stuck, or have questions, please ask them in the #801 thread.

**Comment by JacquesCarette** [comment]
I have reviewed many of the issues, but I started from the oldest ones, so I'm only up to #1415. I'm hoping to get up to current later today.

From my point of view, of the older issues, the ones that I noted down as being things we ought to look at are: #359, #373, #388, #427, #453, #467, #498, #541, #578, #663, #695, #702, #801, #787, #872, #946, #1018, #1025, #1030, #1047, #1076, #1116, #1142, #1146, #1153, #1172, #1173, #1196, #1206, #1210, #1235, #1280, #1311, #1342, #1356, #1415.

**Comment by JacquesCarette** [comment]
[I will come back to this issue when I've had time to go through all the activity]

**Comment by JacquesCarette** [comment]
And my second batch: #1432, #1476. #1500, #1504, #1506, #1524, #1538, #1541, #1578, #1599, #1602, #1605, #1637, #1647, #1658, #1661, #1671, #1697, #1732, #1778, #1780, #1795, #1804, #1809, #1814, #1848, #1853, #1916, #1918, and #1973 are all things worthy of looking in to.

**Comment by JacquesCarette** [comment]
Looking at the issues that @muhammadaliog3 picked, starting from the top. Implementing better constraints is definitely worthwhile, but the analysis is also correct: fixing some will involve major design. So probably not the best issues to start with - hopefully in a couple of weeks though.  Best way to proceed there would be as I commented elsewhere: analyse, suggest design in an issue, get design approval, then implement. #872 is easier to just 'do', but since it will touch things that are likely to also change, it's best to wait [sorry]. #806 is actually going to be surprisingly harder than it sounds! Go ahead and analyse it, it will be an interesting learning experience. #1048 is much lower priority, and perhaps not even the right way to do things.  #1476, #1570, #1809 and #1732 should all be reasonably straightforward - give them a go!

**Comment by JacquesCarette** [comment]
For @Nathaniel-Hu 's list: #946 is definitely something we want - but tricky. As I said above, that one will need to go through the analysis, propose design, approval, implement cycle. #806 was analyzed above. You're working on #801 now. #1018 should be straightforward enough, give it a try. #2088 should be tackled 'later', as it will indeed involve quite a bit of design. #1998 is a tad bit harder than it sounds, but would be worth trying while @bmaclach is still around. #1973 would be a fun DevOps thing to do. #1916 would be very valuable; I think there are some checks already, but there could be more. #1879 might be tricky, another one to understand while @bmaclach is still around to ask questions of. I think #1504 is somewhat lower priority right now.

**Comment by JacquesCarette** [comment]
I'm going to close this issue now, because all the work has been done. However, feel free to use it to ask questions about any of my comments above. For specific issues though, I would prefer if you wrote on the issue itself.

**Comment by smiths** [comment]
I think this was a useful issue.  We now have prioritized lists of issues to go through.  @muhammadaliog3 and @Nathaniel-Hu, as you work through your list, following the advice of @JacquesCarette, please assign the issue you are working on to yourself.  It is good to have "official" ownership of issues.

**Comment by smiths** [comment]
@Nathaniel-Hu and @muhammadaliog3, we have added another issue to your to do list: #2123.  This shouldn't be too difficult an issue to close, and it should be satisfying, since it will increase the number of generated artifacts.  :smile:

**Comment by smiths** [comment]
@Nathaniel-Hu and @muhammadaliog3, we have added another issue to your to do list: #1010.  This should be a straightforward issue to close.

If you feel like you are waiting for @JacquesCarette or @smiths for feedback on a current issue, you should assign yourself another issue and work on it in parallel.  That way we will see progress from you, even when we are too busy to reply to your questions right away.

**Comment by smiths** [comment]
@Nathaniel-Hu and @muhammadaliog3, @bmaclach has created another issue for you to consider: #2180.
